lucene_qat_wrapper/lucene_with_qat.diff

```diff
@@ -0,0 +1,588 @@
++
++    byte[] compressed;
++    int directBufferSize = 655360;
++
++    QatDecompressor() {
++      compressed = new byte[0];
++    }
++
++    QatDecompressor(int directBufferSize) {
++      this.directBufferSize = directBufferSize;
++      compressed = new byte[0];
++    }
++
++    @Override
++    public void decompress(DataInput in, int originalLength, int offset, int length, BytesRef bytes) throws IOException {
++      assert offset + length <= originalLength;
++      if (length == 0) {
++        bytes.length = 0;
++        return;
++      }
++      final int compressedLength = in.readVInt();
++      // pad with extra "dummy byte": see javadocs for using Inflater(true)
++      // we do it for compliance, but it's unnecessary for years in zlib.
++      final int paddedLength = compressedLength;
++      compressed = ArrayUtil.grow(compressed, paddedLength + 1);
++      in.readBytes(compressed, 0, compressedLength);
++      compressed[compressedLength] = 0; // explicitly set dummy byte to 0
++
++      final QatDecompressorJNI decompressor = new QatDecompressorJNI();
++
++      try {
++        // extra "dummy byte"
++        decompressor.setInput(compressed, 0, paddedLength);
++        bytes.offset = bytes.length = 0;
++        bytes.bytes = ArrayUtil.grow(bytes.bytes, originalLength);
++        try {
++          bytes.length = decompressor.decompress(bytes.bytes, bytes.length, originalLength);
++        } catch (Error e) {
++          String s = e.getMessage();
++          System.out.println(e.getMessage());
++        }
++
++        if (!decompressor.finished()) {
++          throw new CorruptIndexException("Invalid decoder state in QAT decompressor: needsInput=" + decompressor.needsInput()
++              + ", needsDict=" + decompressor.needsDictionary(), in);
++        }
++      } finally {
++        decompressor.end();
++      }
++      if (bytes.length != originalLength) {
++        throw new CorruptIndexException("Lengths mismatch in QAT decompressor: " + bytes.length + " != " + originalLength, in);
++      }
++      bytes.offset = offset;
++      bytes.length = length;
++    }
++
++    @Override
++    public Decompressor clone() {
++      return new QatDecompressor();
++    }
++
++  }
++
++  private static class QatCompressor extends Compressor {
++
++    final QatCompressorJNI compressor;
++    byte[] compressed;
++    boolean closed;
++
++    QatCompressor() {
++      compressor = new QatCompressorJNI();
++      compressed = new byte[64];
++    }
++
++    @Override
++    public void compress(byte[] bytes, int off, int len, DataOutput out) throws IOException {
++      compressor.reset();
++      compressor.setInput(bytes, off, len);
++      compressor.finish();
++
++      int totalCount = 0;
++      for (; ; ) {
++        final int count = compressor.compress(compressed, totalCount, compressed.length - totalCount);
++        totalCount += count;
++        assert totalCount <= compressed.length;
```

Comment:
Why do we need this assert? What's case for compressed array length bigger than total count?